### PR TITLE
Add support for Integral and Piecewise in Equation

### PR DESCRIPTION
**Pull request by schymans**
This solves #55 and adds support for `Integral` in `Equation`:
```python
    class valid_general_integral(Equation):
        expr = Eq(demo_d, Integral(demo_v, demo_t))
    class valid_specific_integral(Equation):
        expr = Eq(demo_d, Integral(demo_v, (demo_t, 0, demo_t1)))
```

### CODE REVIEW

**Comment by schymans** [comment]
Travis returns a strange failure, syntax errors in pytest? On my local system, `python setup.py test` does not return any errors.

**Comment by schymans** [comment]
And the Pythong 3.6 failure relates to pydocstyle failures for files that were not changed at all. Not sure what is going on.

File: tests/test_equations.py [hunk 1]
```diff
@@ -107,6 +122,49 @@ def test_integral():
     )
 
 
+def test_Integral():
```
Comment (jirikuncar) [inline tests/test_equations.py]:
```suggestion
def test_integral():
```

File: tests/test_equations.py [hunk 2]
```diff
@@ -107,6 +122,49 @@ def test_integral():
...
+            expr = Eq(demo_d, Integral(demo_v, (demo_t, 1, demo_t1)))
+
+
+def test_log():
+    """Test that log(unit) is dimensionless, as integrate(1/x, x) = log(x)."""
+    class valid_log(Equation):
+        expr = Eq(demo_d, demo_d1 * log(demo_t))
+
+
+def test_Piecewise():
```
Comment (jirikuncar) [inline tests/test_equations.py]:
```suggestion
def test_piecewise():
```

File: tests/test_equations.py (hunk 1 above)
Comment (schymans) [reply tests/test_equations.py]: Thanks, I also changed test_Piecewise() to test_piecewise()

File: essm/variables/_core.py [hunk 3]
```diff
@@ -134,6 +136,15 @@ def get_dimensional_expr(expr):
             for independent, count in expr.variable_count:
                 dim /= Variable.get_dimensional_expr(independent) ** count
             return dim
+        elif isinstance(expr, Integral):
+            dim = Variable.get_dimensional_expr(expr.args[0] *
+                                                expr.args[1][0])
+            return dim
+        elif isinstance(expr, Piecewise):
+            dim = \
```
Comment (jirikuncar) [inline essm/variables/_core.py]: do we need the \ here?

```
dim = Variable.get_dimensional_expr(
  sum(...)
)
```

Comment (schymans) [reply essm/variables/_core.py]: No, thanks. Just removed it.

File: essm/variables/_core.py [hunk 4]
```diff
@@ -202,6 +215,18 @@ def collect_factor_and_basedimension(expr):
                 factor /= ifactor**count
                 dim /= idim**count
             return factor, dim
+        elif isinstance(expr, Integral):
+            # Test that integration limits have the same units
+            Variable.collect_factor_and_basedimension(sum(expr.args[1]))
```
Comment (jirikuncar) [inline essm/variables/_core.py]: does it raise an exception here?

```
try:
   Variable.collect_factor_and_basedimension(sum(expr.args[1]))
except ValueError:
   raise ValueError("Integration limits must have the same units ({expr}).".format(expr=expr))
```

File: essm/variables/_core.py [hunk 5]
```diff
@@ -202,6 +215,18 @@ def collect_factor_and_basedimension(expr):
...
+            factor, dim = \
```
Comment (jirikuncar) [inline essm/variables/_core.py:224]:
```suggestion
            factor, dim = Variable.collect_factor_and_basedimension(
```

File: tests/test_equations.py [hunk 6]
```diff
@@ -84,6 +97,8 @@ class valid_units(Equation):
 
         expr = Eq(demo_v, Derivative(demo_d, demo_fall.definition.t))
 
+    assert(derive_baseunit(valid_units.rhs) == meter / second)
```
Comment (jirikuncar) [inline tests/test_equations.py]:
```suggestion
    assert derive_baseunit(valid_units.rhs) == meter / second
```

File: essm/variables/_core.py (hunk 4 above)
Comment (schymans) [reply essm/variables/_core.py]: I will add a test and check.

File: essm/_generator.py [hunk 7]
```diff
@@ -129,9 +129,9 @@ class VariableWriter(object):
     """Generate Variable definitions.
 
     Example:
-
```
Comment (jirikuncar) [inline essm/_generator.py:132]: @schymans is it necessary to change the formatting here?

File: essm/variables/_core.py (hunk 4 above)
Comment (schymans) [reply essm/variables/_core.py]: OK, I added the try-except block with a slightly shortened error message.

File: essm/_generator.py (hunk 7 above)
Comment (schymans) [reply essm/_generator.py:132]: Travis complained that there should be no blank line below a new section. Not sure if this classifies as a new section.

Comment (schymans) [reply essm/_generator.py:132]: OK, thanks for fixing it!

**Comment by schymans** [comment]
Thanks for fixing some of the issues, Jiri! Unfortunately, pytest for Python 2.7 is still failing. Any ideas?

**Review by jirikuncar — APPROVED** [review]

**Pull request merged**